src/program.hpp

```diff
@@ -163,6 +164,121 @@ enum class build_operation
 
 using cvk_program_callback = void (*)(cl_program, void*);
 
+using cvk_spec_constant_map = std::map<uint32_t, uint32_t>;
+
+struct cvk_program;
+
+class cvk_entry_point {
+public:
+    cvk_entry_point(VkDevice dev, cvk_program* program, const std::string& name)
+        : m_device(dev), m_program(program), m_name(name),
+          m_pod_descriptor_type(VK_DESCRIPTOR_TYPE_MAX_ENUM),
+          m_pod_buffer_size(0u), m_has_pod_arguments(false),
+          m_descriptor_pool(VK_NULL_HANDLE), m_pipeline_layout(VK_NULL_HANDLE),
+          m_pipeline_cache(VK_NULL_HANDLE) {}
+
+    ~cvk_entry_point() {
+        for (auto pipeline : m_pipelines) {
+            cvk_info("destroying pipeline %p for kernel %s", pipeline.second,
+                     m_name.c_str());
+            vkDestroyPipeline(m_device, pipeline.second, nullptr);
+        }
+        if (m_pipeline_cache != VK_NULL_HANDLE) {
+            vkDestroyPipelineCache(m_device, m_pipeline_cache, nullptr);
+        }
+        if (m_descriptor_pool != VK_NULL_HANDLE) {
+            vkDestroyDescriptorPool(m_device, m_descriptor_pool, nullptr);
+        }
+        if (m_pipeline_layout != VK_NULL_HANDLE) {
+            vkDestroyPipelineLayout(m_device, m_pipeline_layout, nullptr);
+        }
+        for (auto layout : m_descriptor_set_layouts) {
+            vkDestroyDescriptorSetLayout(m_device, layout, nullptr);
+        }
+    }
+
+    CHECK_RETURN cl_int init();
+
+    CHECK_RETURN VkPipeline
+    create_pipeline(const cvk_spec_constant_map& spec_constants);
+
+    const std::vector<kernel_argument>& args() const { return m_args; }
+
+    VkDescriptorPool descriptor_pool() const { return m_descriptor_pool; }
```

Comment:
The descriptor pool should not be externally accessible (it is externally-synchronised). This makes it much harder to respect the Vulkan requirements.